ecosystem/indexer-grpc/indexer-grpc-data-service-v2/src/service.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::connection_manager::ConnectionManager;
+use anyhow::Result;
+use aptos_indexer_grpc_utils::timestamp_now_proto;
+use aptos_protos::indexer::v1::{
+    data_service_server::DataService, ping_data_service_response::Info, raw_data_server::RawData,
+    GetTransactionsRequest, HistoricalDataServiceInfo, LiveDataServiceInfo, PingDataServiceRequest,
+    PingDataServiceResponse, StreamInfo, TransactionsResponse,
+};
+use futures::{Stream, StreamExt};
+use std::{pin::Pin, sync::Arc};
+use tokio::sync::mpsc::{channel, Sender};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{Request, Response, Status};
+
+type ResponseStream = Pin<Box<dyn Stream<Item = Result<TransactionsResponse, Status>> + Send>>;
+
+// Note: We still allow starting both services together, so people don't have to rely on
+// GrpcManager for routing, and it's also make it easier to run in testing environment.
+pub struct DataServiceWrapperWrapper {
+    live_data_service: Option<DataServiceWrapper>,
+    historical_data_service: Option<DataServiceWrapper>,
+}
+
+impl DataServiceWrapperWrapper {
+    pub fn new(
+        live_data_service: Option<DataServiceWrapper>,
+        historical_data_service: Option<DataServiceWrapper>,
+    ) -> Self {
+        Self {
+            live_data_service,
+            historical_data_service,
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl DataService for DataServiceWrapperWrapper {
+    type GetTransactionsStream = ResponseStream;
+
+    async fn get_transactions(
+        &self,
+        req: Request<GetTransactionsRequest>,
+    ) -> Result<Response<Self::GetTransactionsStream>, Status> {
+        if let Some(live_data_service) = self.live_data_service.as_ref() {
+            if let Some(historical_data_service) = self.historical_data_service.as_ref() {
+                let request = req.into_inner();
+                let mut stream = live_data_service
```

Comment:
I wonder if we should make this peek at the TonicStatus level since all we do here is to find if live data service contains the data.

Reply:
I'm not sure what you exactly mean, can you give an example?

Reply:
when livedataservice doesn't have the data requested, return `tonic_status::NotFound`.

Reply:
this change may require clients to handle the errors differently. closed.